### DISCUSSION TITLE
Spacemesh

### DISCUSSION THREAD

**Discussion by sa8** [consensus-shipyard/consensuslab#110] [Reading group]
## Protocol

- Multiple bocks mined in parallel (DAG-structure)
    - honestly generated blocks always become valid
    - block publication times are deterministic rather than stochastic: a layer of blocks is published at fixed intervals
    - Every block explicitly contains its layer index—the layer at which the block claims to have been published—and a unique block ID
    - blocks are ordered by layer, then by block ID inside a layer (not sure if grindable or not)
- Protocol operates in two parts: 
    - Tortoise: “slow but sure” eventual consensus. A block includes, in addition to the transaction data, a vote for every previous block.
        - Miners decide how to vote about a block B in the past by counting the votes for and against it (any block generated after B that didn’t vote for it is considered to vote against it)
        - Works well for blocks far in the past but not recent blocks (not enough blocks).
        - instead of always voting using the majority vote, the honest miner considers two different cases: (1) If the vote margin is “large”, the miner votes according to the majority. (2) If the vote margin is “small”, the miner instead votes according to the result of the coin flip.
            - threshold for “large” set by bounding how much honest miners can disagree. Ensures that two honest miners cannot see a “large” voting margin in opposing directions. At the same time, if all honest parties agree on the validity of B and vote appropriately, in the next round all honest parties should have a vote margin that is “large”.
            - “weak” coin is sufficient: as long as the adversary can’t predict the result with more than constant probability, and can’t bias it “too far”,  eventual convergence is guarantees (the expected time it takes to reach consensus does grow as the coin “weakens”, however).
    - Hare: allows very fast convergence. fast, permissionless byzantine agreement protocol.
        - executed after each layer is published, and its output is the set of valid blocks for that layer.
        - all honest miners use the output of the hare protocol when determining how to “vote” in the Tortoise election for recent layers, and use the “majority rule” for older layers.
        - The Hare protocol runs over the gossip network, but only the results of the protocol—as reflected in the miners’ votes—are recorded on the blockmesh. The protocol execution itself does not need to be stored.
        - every block that is rejected by all honest parties will not appear in the output (e.g., blocks that are extremely late will always be considered invalid), while every block that is accepted by all honest parties will always appear in the output (so honest blocks will always be valid).
        - The Hare protocol is based on https://eprint.iacr.org/2018/1028.pdf but for a set of blocks (all the blocks in the layer).
          - A subset is elected in each round (this reduces communication cost). 

### Security guarantees
- secure for f<1/3 + synchronous network (healing in period of asynchrony)
- new users joining the network can agree on the “correct” state as long as they can communicate with a single honest miner
- consensus is immediate if the system is not under attack
- protocol can “self-heal” from an arbitrary violation of our security assumptions, i.e., honest parties will converge to consensus from any initial condition, once the security assumptions are satisfied again. -> equire a period of “relative” quiet—i.e., a period during which the adversary attacks with q << 1/3 of the spacetime resources
- incentive-compatible

## Pros and cons
- Hare is a complex protocol, I am not convinced it scales well (despite the sampling).
- Tortoise seems nice but slow.

**Comment by xuechao2** [comment]
Questions:
1. Is the coin used when the vote margin is small a common coin?
2. Does Hare output the same set of blocks to every honest node? If so, why we still need Tortoise?
3. What is the relationship between the output of Hare and output of Tortoise for one layer? The former is a subset of the latter?

**Reply by sa8** [reply]
1. Yes.
2. If I understand correctly yes and Tortoise helps with the blocks that, for example, arrive late.
3. If I understand correctly yes. Tortoise votes on the blocks output by Hare.

**Reply by sa8** [reply]
Follow-up question from @xuechao2: why do we need to add Tortoise, why isn't Hare sufficient?

**Reply by sa8** [reply]
Hare could fail, for example due to low probability events and as a result honest parties may not agree on the validity of the blocks in layer i. This is where the Tortoise protocol is useful.
